core/controller/src/main/scala/org/apache/openwhisk/core/loadBalancer/FPCPoolBalancer.scala

```diff
@@ -0,0 +1,727 @@
+   **/
+  override def invokerHealth(): Future[IndexedSeq[InvokerHealth]] = {
+    etcdClient.getPrefix(s"${InvokerKeys.prefix}/").map { res =>
+      val healthsFromEtcd = res.getKvsList.asScala.map { kv =>
+        val (memory, busyMemory, status, tags, dedicatedNamespaces) = InvokerResourceMessage
+          .parse(kv.getValue.toString(StandardCharsets.UTF_8))
+          .map { resourceMessage =>
+            val status = resourceMessage.status match {
+              case Healthy.asString   => Healthy
+              case Unhealthy.asString => Unhealthy
+              case Offline.asString   => Offline
+            }
+            (
+              resourceMessage.freeMemory.MB,
+              resourceMessage.busyMemory.MB,
+              status,
+              resourceMessage.tags,
+              resourceMessage.dedicatedNamespaces)
+          }
+          .get
+        val temporalId = InvokerKeys.getInstanceId(kv.getKey.toString(StandardCharsets.UTF_8))
+        val invoker = temporalId.copy(
+          userMemory = memory,
+          busyMemory = Some(busyMemory),
+          tags = tags,
+          dedicatedNamespaces = dedicatedNamespaces)
+
+        new InvokerHealth(invoker, status)
+      }.toIndexedSeq
+      val missingHealths =
+        if (healthsFromEtcd.isEmpty) Set.empty[InvokerHealth]
+        else
+          ((0 to healthsFromEtcd.maxBy(_.id.toInt).id.toInt).toSet -- healthsFromEtcd.map(_.id.toInt))
+            .map(id => new InvokerHealth(InvokerInstanceId(id, Some(id.toString), userMemory = 0 MB), Offline))
+      (healthsFromEtcd ++ missingHealths) sortBy (_.id.toInt)
+    }
+  }
+
+  /** Gets the number of in-flight activations for a specific user. */
+  override def activeActivationsFor(namespace: UUID): Future[Int] =
+    Future.successful(activationsPerNamespace.get(namespace).map(_.intValue()).getOrElse(0))
+
+  /** Gets the number of in-flight activations for a specific controller. */
+  override def activeActivationsByController(controller: String): Future[Int] =
+    Future.successful(activationsPerController.get(ControllerInstanceId(controller)).map(_.intValue()).getOrElse(0))
+
+  /** Gets the in-flight activations */
+  override def activeActivationsByController: Future[List[ActivationId]] =
+    Future.successful(activationSlots.keySet.toList)
+
+  /** Gets the number of in-flight activations for a specific invoker. */
+  override def activeActivationsByInvoker(invoker: String): Future[Int] = Future.successful(0)
+
+  /** Gets the number of in-flight activations in the system. */
+  override def totalActiveActivations: Future[Int] = Future.successful(totalActivations.intValue())
+
+  /** Gets the throttling for given action. */
+  override def checkThrottle(namespace: EntityPath, fqn: String): Boolean = {
+
+    /**
+     * Note! The throttle key is assumed to exist unconditionally and is not limited to throttle if not present.
```

Comment:
Throttling works as follows.
Now the unit of throttling is the number of containers.
And there are two kinds of throttling, 1. `namespace throttling` and 2. `action throttling` with one configuration, `concurrent limit`.
Both throttlings are enabled under the following scenario.

Let's say the concurrent limit of a namespace is 30.
1. When there are 30 containers running in the namespace, the `namespace throttling` is enabled.
2. When the `namespace throttling` is enabled, no more new type of action is accepted.
   1. For example, if `actionA` is invoked and there are 30 containers for `actionA` only activations for `actionA` are allowed.
   2. When another `actionB` is invoked, it is rejected with 429 too many requests.
   3. This is because it's not possible to create more containers in the namespace
   4. When there are two actions `actionA` and `actionB` are running, activations for only those two actions are accepted in the same way.
3. If the workload keeps increasing, activations for `actionA` will be piled up in a queue because there are already 30 containers running and no more containers can be created.
4. If too many activations are buffered in a queue, `action throttling` is enabled. This means that the current containers cannot handle all activations in time. This threshold is configurable.
   1. There are two action throttling configurations, [max retention size](https://github.com/apache/openwhisk/blob/master/core/scheduler/src/main/resources/application.conf#L70), and [throttling fraction](https://github.com/apache/openwhisk/blob/master/core/scheduler/src/main/resources/application.conf#L72).
   2. If there are more activations in a queue than max retention size, action throttling is enabled.
   3. If there are fewer activations than `max-retention-size * throttling fraction`), action throttling is disabled.
   (This is to prevent too frequent throttling changes.)
5. So when a new activation comes, the load balancer checks whether the activation is throttled or not in this way.
   1. Check if `action throttling` is enabled, if yes reject the requests.
   2. If `action throttling` is not enabled, accept the requests.
   3. If `action throttling information` does not exist, checks `namespace throttling`. It means no queue is created for the action.
   4. If `namespace throttling` is enabled, reject the activation. This is because there is no container running and the namespace is already throttled.
   5. If `namespace throttling` is disabled or no information exists, it accepts the activation.

Reply:
For action throttling, is there no longer any concept of invocations per minute throttling? per minute throttling was useful for user applications to throttle to protect their downstream services that their action calls. Now it seems like there's no way to throttle based on what the action application can handle, but is strictly only for system throttling to protect openwhisk

Reply:
Yes. there is no such thing.
If that is necessary, we can add a new feature to the scheduler part. I believe we can control the rate activations are pulled out of a queue.
And if it is critical and urgent, we can indirectly limit invocations with the number of containers with the assumption that one action's execution time does not drastically vary.

Reply:
I'm really doing a deep dive into the code and I'm a little confused on namespace throttling and how you prevent starvation of other actions within a namespace. As you explain, if `actionA` has all of the containers for a namespace and is at the limit but the namespace also wants to run some of `actionB` they just get throttled? If `actionA` sends consistent load to keep up with the container limit for the namespace but not backlog the queue to turn on action throttling, how would `actionB` ever get to run any instances? On the old scheduling algorithm it's just random based on how many activations in flight so even if you're hitting the concurrent throttle, you don't have to worry about one action completely starving another action in the namespace. Am I understanding this correctly?

Reply:
I think in such a case, it is natural to increase the namespace limit.
If someone regularly invokes an action utilizing all the resources and tries to run another one, and if both of the actions are supposed to run well he needs a higher resource limit. In the above case, there are not enough resources and I feel it's not that weird to get a `429 Too many requests` response.

When at least one container is created for `actionB` before the namespace limit is reached, `actionB` can still run as expected even after the namespace is throttled. However, there are not enough containers and activations will get piled up in queues, and eventually, the system will reject the requests with 429 responses if requests keep incoming at the same rate.

To be able to run both of them well, it is necessary to increase the limit to accommodate all of them.
Even if we increase the limit, it would be fine because the system doesn't create more containers than required.